File: LeetCodes/n-th-tribonacci-number/1-30-2023, 10_34_59 AM/n-th-tribonacci-number.cpp
// Link: https://leetcode.com/problems/n-th-tribonacci-number

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: n-th-tribonacci-number
 *  RUNTIME: 0 ms
 *  MEMORY: 6 MB
 *  DATE: 1-30-2023, 10:34:59 AM
 *
 */

// Solution:

class Solution {
public:
    int tribonacci(int n) {
        if(n == 0) return 0;
        if(n == 1 || n == 2) return 1;

        int prev3 = 0, prev2 = 1, prev1 = 1, curr;
        for(int i = 3; i <= n; i++) {
            curr = prev3 + prev2 + prev1;
            prev3 = prev2;
            prev2 = prev1;
            prev1 = curr;
        }
        return curr;
    }
};

File: LeetCodes/n-th-tribonacci-number/4-30-2024, 10_07_32 AM/n-th-tribonacci-number.java
// Link: https://leetcode.com/problems/n-th-tribonacci-number

/** Submission Info:
 *
 *  AUTHOR: the_247 | Raj Patel
 *  TITLE: n-th-tribonacci-number
 *  RUNTIME: 0 ms
 *  MEMORY: 40.1 MB
 *  DATE: 4-30-2024, 10:07:32 AM
 *
 */

// Solution:

class Solution {
    public int tribonacci(int n) {
        if(n < 2) return n;
        int a = 0, b = 1, c = 1;

        for(int i=3; i<=n; i++) {
            int res = a + b + c;
            a = b;
            b = c;
            c = res;
        }
        return c;
    }
}
